codegen/src/main/resources/software/amazon/awssdk/codegen/rules2/RuleUrl.java.resource

```diff
@@ -85,4 +70,58 @@ public final class RuleUrl {
 
         return new RuleUrl(parsed.getProtocol(), parsed.getAuthority(), path, normalizedPath, isIpAddr);
     }
+
+    private static boolean isIpAddr(String host) {
+        int from = 0;
+        int segments = 0;
+        boolean done = false;
+        while (!done) {
+            int index = host.indexOf('.', from);
+            if (index == -1) {
+                if (segments != 3) {
+                    // E.g., 123.com
+                    return false;
+                }
+                index = host.length();
+                done = true;
+            } else if (segments == 3) {
+                // E.g., 1.2.3.4.5
+                return false;
+            }
+            int length = index - from;
+            if (length == 1) {
+                char ch0 = host.charAt(from);
+                if (ch0 < '0' || ch0 > '9') {
+                    return false;
+                }
+            } else if (length == 2) {
+                char ch0 = host.charAt(from);
+                char ch1 = host.charAt(from + 1);
+                if ((ch0 < '0' || ch0 > '9') || (ch1 < '0' || ch1 > '9')) {
+                    return false;
+                }
+            } else if (length == 3) {
+                char ch0 = host.charAt(from);
+                char ch1 = host.charAt(from + 1);
+                char ch2 = host.charAt(from + 2);
+                if ((ch0 < '0' || ch0 > '9')
+                        || (ch1 < '0' || ch1 > '9')
+                        || (ch2 < '0' || ch2 > '9')
+                ) {
+                    return false;
+                }
+                int value = (host.charAt(from) - '0') * 100
+                        + (host.charAt(from + 1) - '0') * 10
+                        + (host.charAt(from + 2) - '0');
```

Comment:
nit: reuse ch0, ch1, ch2 instead of doing another charAt.

Reply:
🤦 Not sure how I missed that.